src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs

```diff
@@ -138,27 +159,31 @@ public static class PackageDownloader
                 throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, Strings.DownloadResourceNotFound, sourceRepository.PackageSource.Source));
             }
 
-            var downloadResourceResult
+            DownloadResourceResult result = null;
+
+            token.ThrowIfCancellationRequested();
+
+            result
                 = await downloadResource.GetDownloadResourceResultAsync(packageIdentity, settings, token);
 
-            if (downloadResourceResult == null)
+            if (result == null)
             {
-                throw new InvalidOperationException(string.Format(
+                throw new FatalProtocolException(string.Format(
```

Comment:
Good we should validate it. Invalid operation exception is the devil

Sent from my Windows Phone
________________________________
From: Justin Emgarten<mailto:notifications@github.com>
Sent: ‎1/‎16/‎2016 11:26 AM
To: NuGet/NuGet.Client<mailto:NuGet.Client@noreply.github.com>
Cc: Yishai Galatzer<mailto:yigalatz@microsoft.com>
Subject: Re: [NuGet.Client] Add retry logic for failed downloads because of broken TCP connections (#202)


In src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs<https://github.com/NuGet/NuGet.Client/pull/202#discussion_r49935990>:

>              {
> -                throw new InvalidOperationException(string.Format(
> +                throw new FatalProtocolException(string.Format(


Changing the exception type here my affect how PS, VS UI, and VS API handle this.

—
Reply to this email directly or view it on GitHub<https://github.com/NuGet/NuGet.Client/pull/202/files#r49935990>.